docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -3,43 +3,134 @@
+and disseminating transactions is maintaining the order in which transactions arrive to the nodes,
+that is, a FIFO ordering. However, ensuring a strict transmission order over the network proves
+challenging due to inherent characteristics of the underlying communication protocols that causes
+message delays and potential reordering. Consequently, while Tendermint Core and CometBFT
+applications have always assumed this supposed ordering, the FIFO-ness of transactions is not
+guaranteed and is offered only as a best effort.
+
+Beyond the apparent FIFO sequencing, transactions in the mempool are treated equally, meaning that
+they are not discriminated as to which are disseminated first or which transactions the mempool
+offers to the proposer when creating the next block. In practice, however, not all transactions have
+the same importance for the application logic, especially when it comes to latency requirements.
+Depending on the application, we may think on countless categories of transactions based on their
+importance and requirements, spanning from IBC messages, transactions for exchanges, for smart
+contract execution, for smart contract deployment, grouped by SDK modules, and so on. Even
+transactions prioritized by economic incentives could be given a preferential treatment. Or big
+transactions, regardless of their nature, could be categorized as low priority, to mitigate
+potential attacks on the mempool.
+
+The goal of this document is thus to propose a mechanism enabling the mempool to prioritize
+transactions by *classes*, for processing and dissemination, directly impacting block creation. In
+IP networking terminology, this is known as Quality of Service (QoS). By providing certain QoS
+guarantees, developers will be able to more easily estimate when transactions will be processed and
+included in a block. 
+
+In practical terms, we envision an implementation of the transaction class abstraction as *mempool
+lanes*. The application will be allowed to split the mempool transaction space into a hierarchy of
+lanes, with each lane operating as an independent mempool. At the same time, all of them need to be
+coordinated to ensure the delivery of the desired levels of QoS.
+
+Note that improving the dissemination protocol to reduce bandwidth and/or latency is a separate
+concern and falls outside the scope of this proposal.
+
+## Properties
+
+Before jumping into the design of the proposal, we define more formally the properties supported by
+the current implementation of the mempool. Then we state what properties the new mempool should
+offer to guarantee the desired QoS.
+
+The following definition is common to all properties. 
+
+:memo: _Definition_: Given any two different transactions `tx1` and `tx2`, we say that `tx1` is
+*processed and disseminated before* `tx2`, when:
+- `tx1` is reaped from the mempool to form a block proposal before `tx2`,
+- `tx1` is rechecked before `tx2`, and
```

Comment:
I'm trying to imagine why _recheck_ is important here

Reply:
For FIFO ordering. If tx2 is rechecked before tx1, one of the two or both may become invalid and removed from the mempool. I'm not 100% sure that we need to keep rechecking in the list.